File: archive/p/python/quick_sort.py
import sys


def quick_sort(xs):
    if len(xs) <= 0:
        return []

    left = quick_sort([l for l in xs[1:] if l <= xs[0]])
    right = quick_sort([r for r in xs[1:] if r > xs[0]])
    return left + xs[:1] + right


def input_list(list_str):
    return [int(x.strip(" "), 10) for x in list_str.split(',')]


def exit_with_error():
    print('Usage: please provide a list of at least two integers to sort in the format "1, 2, 3, 4, 5"')
    sys.exit(1)


def main(args):
    try:
        xs = input_list(args[0])
        if len(xs) <= 1:
            exit_with_error()
        print(quick_sort(xs))
    except (IndexError, ValueError):
        exit_with_error()


if __name__ == "__main__":
    main(sys.argv[1:])

File: archive/p/perl/quick-sort.pl
#!/usr/bin/env perl
use v5.42;

use feature qw/keyword_any/;
no warnings 'experimental::keyword_any';

sub usage {
    say 'Usage: please provide a list of at least two integers to sort in the format "1, 2, 3, 4, 5"';
    exit;
}

sub parse_list ($s) {
    return undef unless defined $s;

    my @vals = split /\s*,\s*/, $s;

    return undef if @vals < 2;
    return undef if any { $_ !~ /\A-?\d+\z/ } @vals;

    return [ map 0 + $_, @vals ];
}

sub quick_sort ($a) {
    return $a if @$a <= 1;

    my $pivot = $a->[ @$a >> 1 ];

    my @left  = grep { $_ < $pivot } @$a;
    my @mid   = grep { $_ == $pivot } @$a;
    my @right = grep { $_ > $pivot } @$a;

    return [ @{ quick_sort( \@left ) }, @mid, @{ quick_sort( \@right ) } ];
}

my ($input) = @ARGV;
my $a = parse_list($input) or usage();

$a = quick_sort($a);
say join ', ', @$a;
